proposals/0000-shared-storage-property-wrapper.md

```diff
@@ -0,0 +1,419 @@
+
+```swift
+struct Hud {
+  @Clamped(.init(0...14)) var bar = 5
+  @Clamped(.init(1...9)) var foo = 1
+}
+
+var hud1 = Hud()
+var hud2 = Hud()
+
+// desugars to
+
+struct Hud {
+  shared let bar$shared = RangeStorage(0...14)
+  shared let foo$shared = RangeStorage(1...9)
+  
+  var bar = Clamped(wrappedValue: 5, shared: bar$shared)
+  var foo = Clamped(wrappedValue: 1, shared: foo$shared)
+}
+
+// both Hud's get access to the same $shared properties.
+var hud1 = Hud() 
+var hud2 = Hud() 
+```
+
+### Initialization
+
+Inside the wrapper's initializer, assigning the shared value to the `shared` property is handled by the compiler, so there's not need to explicitly do it. If the initializer itself needs access to storage properties, it should be done through the `shared` parameter. 
+
+```swift
+shared let storage: RangeStorage<Value>
+
+init(wrappedValue: Value, shared: RangeStorage<Value>) {
+  self.value = shared.clamp(wrappedValue)
+
+  // referencing `storage` inside the initializer is not okay since 
+  // it might not have been initialized yet?
+}
+```
+
+At the call site, we can drop the label for the shared argument if there aren't other arguments to be passed (except wrappedValue and projectedValue). 
+
+```swift
+// without the label
+@Clamped(RangeStorage(1...7)) var weekday: Int = 7
+
+// with a third parameter on the initializer... 
+init(wrappedValue: Value, shared: RangeStorage<Value>, defaultValue: Value) { }
+
+// ... we bring back the label
+@Clamped(shared: RangeStorage(1...7), defaultValue: 1) var weekday: Int = 7
+```
+
+Property wrapper authors may also enable passing the values that will be used to initialize the storage instance directly. That can be accomplished by adding a `@shared` attribute on the initializer. 
+
+```swift
+
+struct RangeStorage<Value> {
+  init(_ range: ClosedRange<Value>) { ... } 
+}
+
+/// ...
+// using `@shared` instead of `shared`...
+init(wrappedValue: Value, @shared: RangeStorage<Value>) {}
+
+// ...
+// call site
+@Clamped(1...7) var weekday: Int = 3
+```
+
+The initialization of the storage value itself follows the same principles as static variables: it can't instance variables or methods that depend on `self` being initialized. Though literals and other type variables can be used. 
+
+```swift
+struct RangeStorage {
+  init(_ range: String) { ... } 
+}
+
+struct Container { 
+  @Clampped(1...7) var weekday = 3
+}
+
+// not okay
+struct ContainerB {
+  var minDay: Int
+  var maxDay: Int
+  @Clampped(minDay...maxDay3) var weekday = 3
+}
+```
+
+As mentioned before, the compiler will enforce that the shared property initialization only happens once. For that reason, the `shared` property is required to be initialized from a `shared` parameter. 
```

Comment:
We briefly discussed this, but I think it makes more sense to not require the line `self.storage = storage`, because it enforces the model that the shared storage is already initialized. What we need to do instead is figure out a way to "match up" the shared parameter with the shared storage, but I think it's better to have new syntax or a new rule for that instead of writing it in a way that looks like stored property initialization.